src/main/scala/piconot/internal/Language.scala

```diff
@@ -0,0 +1,269 @@
+        surRule = Surroundings(Anything, Anything, Open, Open)
+      } else if (sur == FRONT_WALL) {
+        surRule = Surroundings(Anything, Anything, Open, Anything)
+      } else if (sur == RIGHT_LEFT_WALL) {
+        surRule = Surroundings(Open, Anything, Open, Open)
+      } else if (sur == RIGHT_FRONT_WALL) {
+        surRule = Surroundings(Open, Anything, Open, Anything)
+      } else if (sur == RIGHT_LEFT_FRONT_WALL) {
+        surRule = Surroundings(Open, Anything, Open, Open)
+      } else if (sur == LEFT_FRONT_WALL) {
+        surRule = Surroundings(Anything, Anything, Open, Open)
+      } else if (sur == RIGHT_OPENING) {
+        surRule = Surroundings(Blocked, Anything, Open, Anything)
+      } else if (sur == LEFT_OPENING) {
+        surRule = Surroundings(Anything, Anything, Open, Blocked)
+      } else if (sur == RIGHT_LEFT_OPENING) {
+        surRule = Surroundings(Blocked, Anything, Open, Blocked)
+      } else if (sur == RIGHT_BACK_OPENING) {
+        surRule = Surroundings(Blocked, Blocked, Open, Anything)
+      } else if (sur == RIGHT_LEFT_BACK_OPENING) {
+        surRule = Surroundings(Blocked, Blocked, Open, Blocked)
+      } else if (sur == LEFT_BACK_OPENING) {
+        surRule = Surroundings(Anything, Blocked, Open, Anything)
+      } else if (sur == BACK_OPENING) {
+        surRule = Surroundings(Anything, Blocked, Open, Anything)
+      }
+    }
+
+    Rule(State(stCounter + ""), surRule, dirRule, State(stCounter + ""))
+  }
+
+  def evalJump(lbl: String) = {
+    var newState = ""
+    labelStates.get(lbl) match {
+      case Some(lbl) => newState = lbl + ""
+      case _         => null
+    }
+    Rule(State(stCounter + ""), Surroundings(Anything, Anything, Anything, Anything), StayHere, State(newState))
+  }
+}
+
+class Rules {}
+case class Label(name: String)(val rules: List[Rules]) extends Rules {}
+case class Face(direction: Direction) extends Rules
+case class Go(surrounding: Surrounding) extends Rules
+// object Go {  
+//     def until(surrounding: Surrounding) = surrounding
+// }
+
+case class Jump(label: String) extends Rules
+
+class Surrounding extends Rules {}
+class Direction extends Rules {}
+object NORTH extends Direction {}
+object EAST extends Direction {}
+object SOUTH extends Direction {}
+object WEST extends Direction {}
+
+object RIGHT_WALL extends Surrounding {}
+object LEFT_WALL extends Surrounding {}
+object FRONT_WALL extends Surrounding {}
+object RIGHT_LEFT_WALL extends Surrounding {}
+object RIGHT_FRONT_WALL extends Surrounding {}
+object RIGHT_LEFT_FRONT_WALL extends Surrounding {}
+object LEFT_FRONT_WALL extends Surrounding {}
+object BACK_OPENING extends Surrounding {}
+object RIGHT_OPENING extends Surrounding {}
+object LEFT_OPENING extends Surrounding {}
+object RIGHT_LEFT_OPENING extends Surrounding {}
+object RIGHT_BACK_OPENING extends Surrounding {}
+object RIGHT_LEFT_BACK_OPENING extends Surrounding {}
+object LEFT_BACK_OPENING extends Surrounding {}
```

Comment:
I agree this seems like a lot of complexity.  It 'd make me consider writing a program to generate these rules since I know I'd typo at least a few of them.